### PR TITLE
feat: add experimental PWA assets generation and injection

### PR DESCRIPTION
**Pull request by userquin**
This PR includes PWA assets generator `0.2.2`, with the following features:
- generate all icon assets from the preset, including apple splash screen icons
- inject icon assets icons in the PWA web manifest icons in dev and build: disabled by default (the plugn will auto-detect if icons present)
- inject favicon, apple touch icon and apple touch startup image links in the html head: enabled by default
- inject PWA web manifest `color_theme` in the html head meta `theme-color`: enabled by default
- allow inlined and external pwa assets configuration: external configuration has preference over inlined
- support for caching resources in dev: using 304 when already requested (using `id-modified-since` header) and cache invalidation on configuration change
- pwa assets config file reload support in dev: once configuration reloaded and cache cleared, the client will refresh the page
- expose 2 new virtual modules for integrations: `virtual:pwa-assets/head` and `virtual:pwa-assets/icons`
- expose internal pwa assets generator in the api: maybe required for some integration

You can run the `assets-generator` example in dev or building and preview:
- from root run `pnpm install && pnpm build`
- change to `assets-generator` example: `cd examples/assets-generator`
- run the dev script (`pnpm dev`) or build + preview  (`pnpm build && pnpm preview`)

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -194,10 +191,42 @@ export interface VitePWAOptions {
   /**
    * PWA assets generation and injection.
    *
+   * If enabled via boolean value, the plugin will search for the pwa assets generator files in the root directory of your project:
+   * - pwa-assets.config.js
```
Comment (userquin) [inline src/types.ts:195]: add .ts and review .mts and .cts

File: src/types.ts [hunk 2]
```diff
@@ -194,10 +191,42 @@ export interface VitePWAOptions {
...
-   */
-  assets?: false | PWAAssets
+   * @see https://vite-pwa-org.netlify.app/assets-generator/cli.html#configurations
+   */
+  assets?: {
+    /**
+     * PWA assets generation and injection.
+     *
+     * If `true` the plugin will search for the pwa assets generator files in the root directory of your project:
+     * - pwa-assets.config.js
```
Comment (userquin) [inline src/types.ts:212]: add .ts and review .mts and .cts

**Comment by Aslemammad** [comment]
Excited for this PR!

File: src/plugins/main.ts [hunk 3]
```diff
@@ -27,6 +28,21 @@ export function MainPlugin(ctx: PWAPluginContext, api: VitePluginPWAAPI) {
       ctx.viteConfig = config
       ctx.userOptions?.integration?.configureOptions?.(config, ctx.userOptions)
       ctx.options = await resolveOptions(ctx.userOptions, config)
+      try {
+        ctx.assets = await import('../asset-generator').then(({ loadInstructions }) => loadInstructions(ctx))
```
Comment (userquin) [inline src/plugins/main.ts:32]: Try load when assets configured and log when error

File: src/plugins/assets.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        const icon = await assetsGenerator.findIconAsset(url)
+        if (!icon)
+          return next()
+
+        const buffer = await icon[2]
+        res.setHeader('Content-Type', icon[1])
+        res.setHeader('Content-Length', buffer.length)
+        res.setHeader('Cache-Control', 'public,max-age=0,must-revalidate')
+        res.statusCode = 200
+        res.end(buffer)
```
Comment (userquin) [inline src/plugins/assets.ts:38]: make this non blocking

File: src/plugins/assets.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        return []
+      }
+    },
+    configureServer(server) {
+      server.middlewares.use(async (req, res, next) => {
+        const url = req.url
+        if (!url)
+          return next()
+
+        const assetsGenerator = await ctx.assets
```
Comment (userquin) [inline src/plugins/assets.ts:36]: if url isn't ico, svg, png, webp request just `return next()` (don't await and avoid extra lookups for any js/css resource)

File: src/plugins/assets.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
+import type { Plugin } from 'vite'
+
+import type { PWAPluginContext } from '../context'
+import { DEV_RELOAD_PAGE_NAME } from '../constants'
+
+export function AssetsPlugin(ctx: PWAPluginContext) {
+  return <Plugin>{
+    name: 'vite-plugin-pwa:assets',
+    enforce: 'post',
+    transformIndexHtml: {
```
Comment (userquin) [inline src/plugins/assets.ts:10]: only for build: for dev we should use ws ping/pong message

File: src/asset-generator.ts [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+  sources: string[]
+  injectThemeColor: boolean
+  includeHtmlHeadLinks: boolean
+  overrideManifestIcons: boolean
+}
+
+export function loadInstructions(ctx: PWAPluginContext) {
+  return async (): Promise<PWAAssetsGenerator | undefined> => {
+    // if disabled, no assets enabled or no manifest
+    if (!ctx.options.assets || !ctx.options.manifest)
```
Comment (userquin) [inline src/asset-generator.ts:35]: add disable and dev checks

File: src/asset-generator.ts [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+    async checkHotUpdate(file) {
+      const result = assetsContext.sources.includes(file)
+      if (result)
+        await loadContext(assetsContext)
+
+      return result
+    },
+  } satisfies PWAAssetsGenerator
+
+  async function loadContext(assetContext?: AssetsGeneratorContext) {
```
Comment (userquin) [inline src/asset-generator.ts:164]: move this outside

File: src/plugins/build.ts [hunk 9]
```diff
@@ -26,19 +26,29 @@ export function BuildPlugin(ctx: PWAPluginContext) {
...
       return _generateBundle(ctx, bundle)
     },
     closeBundle: {
       sequential: true,
       order: ctx.userOptions?.integration?.closeBundleOrder,
       async handler() {
-        if (!ctx.viteConfig.build.ssr && !ctx.options.disable)
-          await _generateSW(ctx)
+        if (!ctx.viteConfig.build.ssr) {
+          const assetsGenerator = await ctx.assets
```
Comment (userquin) [inline src/plugins/build.ts:45]: add only this also if enabled

**Comment by userquin** [comment]
Moved pwa assets dependency to optional peer, here small repro if someone want to check it https://github.com/userquin/vite-plugin-pwa-assets-generator (local tgz from pwa plugin PR)

**Pull request merged**